Added full generic type definitions, and improved `get` definitions.

freshollie opened the pull request:
This PR makes `OatyArray` fully generic. It does this by ensuring that key definitions have to exist on the objects in the input array, and by ensuring that the `get` method will only accept queries for key and value types which do exist.

This meant removing some of the tests around checking errors when trying to access non existent keys, as they would be thrown at compile time. If you think these should be put back, that's fine. 

This addition also means that autocompletion is greatly improved when interacting `oaty`.

Here is an example of what these definitions give you:

### Typechecked access
```typescript
const myArray = new OatyArray([{ a: "a", b: 1, c: 50}])
myArray.get("d") // ERROR: Argument of type '"d"' is not assignable to parameter of type '"a" | "b" | "c"'
```

```typescript
const myArray = new OatyArray([{ a: "a", b: 1, c: 50}])
myArray.get("a", 10) // ERROR: Argument of type '10' is not assignable to parameter of type 'string'
```

```typescript
const myArray = new OatyArray([{ a: "a", b: 1, c: 50}], { keys: ["c"]});
const values = myArray.get("b", 1); // Argument of type '"b"' is not assignable to parameter of type '"c"'.
```

### Correct return types
```typescript
const myArray = new OatyArray([{ a: "a", b: 1, c: 50}])

const values = myArray.get("b");
values["a"] // ERROR: Element implicitly has an 'any' type because index expression is not of type 'number'
```
```typescript
const myArray = new OatyArray([{ a: "a", b: 1, c: 50}])

const values = myArray.get("b");
values[1][0].d // ERROR: Property 'd' does not exist on type '{ a: string; b: number; c: number; }'
```

### Enforced `keys` input
```typescript
const myArray = new OatyArray([{ a: "a", b: 1, c: 50}], { keys: ["B"]}) // ERROR: Argument of type '{ keys: "B"[]; }' is not assignable to parameter of type 'Options<"a" | "b" | "c">'. Types of property 'keys' are incompatible.
```

freshollie committed:
feat: add definitions for full generic

freshollie committed:
fix: remove optional as not possible

freshollie committed:
fix: put back undefined check

freshollie committed:
fix: added mission overload definition

freshollie committed:
fix: remove check for keys outside not in object

jmsv reviewed (changes requested):
looks good! - could we could get some more unit tests on this to replace the removed ones?

freshollie commented:
Can do. The unit tests I removed were because their usecase is now enforced at compile time, rather than runtime. I can add back the functionality to check at runtime too, if you think it necessary.

j-m reviewed (commented):
I like this. 
I think we should keep the runtime checks for use of the module in JavaScript. 
Now that we have `T,K,V,A` types instead of just `T` I think it would sensible to give them more descriptive names. 
Also, isn't K slightly redundant? I thought object keys were always strings, or can you think of a use-case where they aren't?

freshollie commented:
@j-m keys of objects are not just strings, but string literals, so they need their own type. This is why this change is so effective.

I agree that the Generic typenames are confusing. By convention generic types should be as short as possible. I will try to think of something to make them more clear.

I'll add back the runtime checks.

freshollie commented:
@j-m Example of `K` being more than just `string`

```typescript
const myArray = new OatyArray([{ a: "a", b: 1, c: 50}])

type Keys = typeof myArray.keys
// type Keys = ("a" | "b" | "c")[] | undefined

const keys = myArray.keys;
keys && keys.find(a => a === "hello") // ERROR: This condition will always return 'false' since the types '"a" | "b" | "c"' and '"hello"' have no overlap.
```

freshollie committed:
chore: improve typenames

freshollie commented:
FYI, I've realised part of the requirements is that `keys` can be given without a data array to infer the types from. Gonna make some changes to allow for this behaviour.

freshollie committed:
feat: improved definitions, updated to TS3.8, added test-d definitions

freshollie committed:
chore: change generic name

freshollie committed:
fix: more type fixes

freshollie committed:
fix: fixed missing infered keys from input objects

freshollie committed:
chore: added one more definition test

freshollie committed:
chore: remove whitespace

freshollie committed:
fix: corrected keys

freshollie committed:
fix: better keys value

freshollie committed:
chore: formatting

freshollie commented:
@jmsv @j-m I've updated this PR to conform to the original specification. To do this I had to update typescript to 3.5+, so I used 3.8. I hope this is OK.

I have added [tsd](https://github.com/SamVerschueren/tsd) tests in order to both prove and check this specification of types, which can be found at `test/index.test-d.ts`

I have also made a very small change in the API. From now`keys` property will be equal to the keys of the initial values, unless a subset of those keys is specified in options. I've added tests for this. If someone really doesn't want specific keys to be transposed, then not providing an initial array or use a generic `any` will do the job.

jmsv commented:
looks really good :100: I wanna get this merged but please could you write a quick changelog entry in the readme? will need to be a [major](https://semver.org) version bump to `1.0.0` due to the breaking changes on the `OatyArray` constructor. usage examples in the readme also need updating but I don't mind doing this after merge

freshollie committed:
chore: updated changelog

freshollie commented:
Is this what you were looking for? Do you want me to bump the version?

jmsv commented:
@freshollie perfect thanks! I'll bump the version

jmsv commented:
[published](https://www.npmjs.com/package/oaty/v/1.0.0)! thanks for the work on this @freshollie
